oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java

```diff
@@ -0,0 +1,159 @@
+ */
+public class OAuth2ResourceServerConfigurer<B extends HttpSecurityBuilder<B>> extends
+		AbstractHttpConfigurer<OAuth2ResourceServerConfigurer<B>, B> {
+
+	private AccessTokenFormatsConfigurer accessTokenFormatsConfigurer;
+	private AccessTokenVerifiersConfigurer accessTokenVerifiersConfigurer;
+
+	public OAuth2ResourceServerConfigurer<B> bearerTokenResolver(BearerTokenResolver resolver) {
+		return this;
+	}
+
+	public AccessTokenConfigurer accessToken(OAuth2TokenVerifier... verifiers) {
+		return new AccessTokenConfigurer();
+	}
+
+	public class AccessTokenConfigurer {
+		public AccessTokenVerifiersConfigurer verifiers() {
+			return new AccessTokenVerifiersConfigurer();
+		}
+
+		public AccessTokenFormatsConfigurer formats() {
+			return new AccessTokenFormatsConfigurer();
+		}
+
+		public OAuth2ResourceServerConfigurer<B> and() {
+			return OAuth2ResourceServerConfigurer.this;
+		}
+	}
+
+	public class AccessTokenVerifiersConfigurer {
+		public SignatureVerificationConfigurer signature() {
+			return new SignatureVerificationConfigurer();
+		}
+
+		public EncryptionVerificationConfigurer encryption() {
+			return new EncryptionVerificationConfigurer();
+		}
+
+		public AccessTokenVerifiersConfigurer addVerifier(OAuth2TokenVerifier verifier) {
+			return this;
+		}
+
+		public AccessTokenConfigurer and() {
+			return null;
+		}
+	}
+
+	public class SignatureVerificationConfigurer {
+		public SignatureVerificationConfigurer keys(String uri) {
+			return this;
+		}
+
+		public AccessTokenVerifiersConfigurer and() {
+			return OAuth2ResourceServerConfigurer.this.accessTokenVerifiersConfigurer;
+		}
+	}
+
+	public class EncryptionVerificationConfigurer {
+		public EncryptionVerificationConfigurer keys(String uri) {
+			return this;
+		}
+
+		public AccessTokenVerifiersConfigurer and() {
+			return OAuth2ResourceServerConfigurer.this.accessTokenVerifiersConfigurer;
+		}
+	}
+
+	public class AccessTokenFormatsConfigurer {
+		public OpaqueAccessTokenFormatConfigurer opaque() {
+			return new OpaqueAccessTokenFormatConfigurer();
+		}
+
+		public JwtAccessTokenFormatConfigurer jwt() {
+			return new JwtAccessTokenFormatConfigurer();
+		}
+
+		public AccessTokenConfigurer and() {
+			return null;
+		}
+	}
+
+	public class OpaqueAccessTokenFormatConfigurer {
+		public OpaqueAccessTokenFormatConfigurer processor
+				(OAuth2AccessTokenProcessor<? extends AbstractOAuth2Token> processor) {
+			return this;
+		}
+
+		public AccessTokenFormatsConfigurer and() {
+			return OAuth2ResourceServerConfigurer.this.accessTokenFormatsConfigurer;
+		}
+	}
+
+	public class JwtAccessTokenFormatConfigurer {
+		public JwtAccessTokenFormatConfigurer processor
+				(OAuth2AccessTokenProcessor<Jwt> processor) {
+			return this;
+		}
+
+		public AccessTokenFormatsConfigurer and() {
+			return OAuth2ResourceServerConfigurer.this.accessTokenFormatsConfigurer;
+		}
+	}
+
+	private interface OAuth2AccessTokenProcessor<T extends AbstractOAuth2Token> {
```

Comment:
What is `OAuth2AccessTokenProcessor` used for?

Reply:
Was looking for language in the spec that indicates the work that is done going from a token to a set of claims. Wanted something that made the notion of `JwtDecoder` more generic.

We have this same kind of thing with opaque tokens: Take this opaque token, go to this endpoint, confirm its validity, and (maybe) turn it into a claim set.

Essentially processing is: Take this token and turn it into a claim set.

Reply:
It's still not clear to me why we need this. If the user provided this configuration:

`oauth2().resourceServer().accessToken().jwt().jwkSetUri(jwkSetUri)`

than the `SecurityConfigurer` would initialize a `JwtAccessTokenAuthenticationProvider` and associate it with a `NimbusJwtDecoderJwkSupport` based on this configuration.

The point is that the `JwtDecoder` implementation would be known to the `SecurityConfigurer` based on the configuration.

Make sense?

Reply:
Agreed that we could correctly infer the default configuration from what you have listed.

What I am looking for is the right way for code to specify overrides to the defaults. I remember that you mentioned this as a possibility:

```java
@Bean
public JwtDecoder jwtDecoder() {
    // ...
}

@Bean
public OpaqueTokenIntrospector inspector() {
    // ... or whatever we call the thing that translates an opaque token into an instance of AbstractOAuth2Token
}
```
Is that the approach you would advocate?

If so, I guess what I'm not clear on is why the above way is more desirable than this:

```java
oauth2().resourceServer().accessToken().jwt().processor(decoder());

// ... or

oauth2().resourceServer().accessToken()
    .opaque().processor(introspector()).verifiers(...).and()
    .jwt().processor(decoder()).verifiers(...);
```

To me, with the @Bean way, I'd need to look it up in the documentation whereas this way, I can do tab completion to see what configuration options I have. Here, I can read it like I read the spec "My Jwt Processor is X".